tfjs-backend-webgpu/src/kernels/from_pixels_webgpu.ts

```diff
@@ -26,35 +26,131 @@ export class FromPixelsProgram implements WebGPUProgram {
+
+  inputTexture: GPUTexture = null;
+  lastPixelSize = {width: 0, height: 0};
 
   constructor(outputShape: number[]) {
     this.outputShape = outputShape;
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape);
-
+    const workGroupSize = 256;
     this.userCode = `
-      void main() {
-        ivec3 coords = getOutputCoords();
-        int texR = coords[0];
-        int texC = coords[1];
-        int depth = coords[2];
-        vec2 uv = (vec2(texC, texR) + halfCR) / vec2(outShape.yx);
-
-        vec4 values = texelFetch(A, uv);
-        float value;
-        if (depth == 0) {
-          value = values.r;
-        } else if (depth == 1) {
-          value = values.g;
-        } else if (depth == 2) {
-          value = values.b;
-        } else if (depth == 3) {
-          value = values.a;
-        }
+    layout (local_size_x = ${workGroupSize},
+      local_size_y = 1,
+      local_size_z = 1) in;
+    layout(set = 0, binding = 1, rgba8) uniform readonly image2D srcImage;
+    layout(set = 0, binding = 2) uniform Meta {
+      int size;
+      int width;
+      int numChannels;} outShape;
+
+    ivec3 getOutputCoordsIB() {
+      int index0 = int(gl_GlobalInvocationID[0]) * outShape.numChannels;
+      int d0 = index0 / (outShape.numChannels * outShape.width);
+      index0 -= d0 * (outShape.numChannels * outShape.width);
+      int d1 = index0 / outShape.numChannels;
+      int d2 = index0 - d1 * outShape.numChannels;
+      return ivec3(d0,d1,d2);
+    }
 
-        setOutput(floor(value * 255.0 + 0.5));
+    void main() {
+      ivec3 coords = getOutputCoordsIB();
+      int texR = coords[0];
+      int texC = coords[1];
+      int depth = coords[2];
+      vec4 values = imageLoad(srcImage, ivec2(texC, texR));
+      float value;
+      for(int i = 0; i < outShape.numChannels; i++) {
+        value = values[i];
+        int flatIndex =
+          int(gl_GlobalInvocationID.x) * outShape.numChannels + i;
+        if (flatIndex < outShape.size) {
+          result[flatIndex] = int(floor(255.0 * value));
+        }
       }
+
+    }
     `;
     this.shaderKey = 'fromPixel';
   }
+
+  setProperties(
+      bindGroupLayout: GPUBindGroupLayout, pipeline: GPUComputePipeline) {
+    this.bindGroupLayout = bindGroupLayout;
+    this.pipeline = pipeline;
+  }
+
+  setUniform(device: GPUDevice, uniformData: number[]) {
+    const [uniformBuffer, uniformMapping] =
+        device.createBufferMapped({size: 4 * 3, usage: GPUBufferUsage.UNIFORM});
+    new Uint32Array(uniformMapping).set(uniformData);
+    uniformBuffer.unmap();
+    this.uniform = uniformBuffer;
+  }
+
+  getInputTexture(device: GPUDevice, pixelWidth: number, pixelHeight: number):
+      GPUTexture {
+    if (!this.inputTexture || this.lastPixelSize.width !== pixelWidth ||
+        this.lastPixelSize.height !== pixelHeight) {
+      this.inputTexture = device.createTexture({
```

Comment:
Do we need to care when to dispose this texture?

Reply:
A good question. Per my opnion, this texture won't be reused in another program in this patch original design. Based on this, I'd like to leave the destruct of this inputTexture (and the uniform buffer) for js object lifecycle. WDYT?

Reply:
I don't think you can rely on the js object lifecycle to manage GPU textures - but maybe I'm missing something - cc @kainino0x

Reply:
oops nevermind kai just saw that you already addressed this - reviewable and github reviews don't work together sigh

Reply:
Done. (By adding dispose logic).